components/email_aliases/email_aliases_service.cc

```diff
@@ -30,21 +89,198 @@ void EmailAliasesService::BindInterface(
+                                                         "application/json");
+  verification_simple_url_loader_->DownloadToString(
+      url_loader_factory_.get(),
+      base::BindOnce(&EmailAliasesService::OnRequestSessionResponse,
+                     weak_factory_.GetWeakPtr()),
+      kMaxResponseLength);
+}
+
+void EmailAliasesService::OnRequestSessionResponse(
+    std::optional<std::string> response_body) {
+  if (!response_body) {
+    // No response body, log it and re-request.
+    LOG(ERROR) << "Email Aliases service error: No response body";
+    MaybeRequestSessionAgain();
+    return;
+  }
+  const auto response_body_dict = base::JSONReader::ReadDict(*response_body);
+  if (!response_body_dict) {
+    // Invalid response body, log it and re-request.
+    LOG(ERROR) << "Email Aliases service error: Invalid response body";
+    MaybeRequestSessionAgain();
+    return;
+  }
+  auto error_message = ErrorResponse::FromValue(*response_body_dict);
+  if (error_message) {
+    // An error has been reported by the server, indicating that verification
+    // failed. Log it and notify observers.
+    LOG(ERROR) << "Email Aliases service error: " << error_message->error;
+    session_poll_elapsed_timer_.reset();
+    NotifyObserversAuthStateChanged(
+        mojom::AuthenticationStatus::kUnauthenticated,
+        /*error_message=*/l10n_util::GetStringUTF8(
+            IDS_EMAIL_ALIASES_ERROR_VERIFICATION_FAILED));
+    return;
+  }
+  auto parsed_session = SessionResponse::FromValue(*response_body_dict);
+  if (!parsed_session) {
+    // No error message but unparseable response, log it and re-request.
+    LOG(ERROR) << "Email Aliases service verification error: Parse error but "
+                  "no error message";
+    MaybeRequestSessionAgain();
+    return;
+  }
+  if (!parsed_session->verified || !parsed_session->auth_token) {
+    // Verification still in progress; no auth token yet. Re-request.
+    MaybeRequestSessionAgain();
+    return;
+  }
+  // Success; set the auth token and notify observers.
+  auth_token_ = *parsed_session->auth_token;
+  session_poll_elapsed_timer_.reset();
+  NotifyObserversAuthStateChanged(mojom::AuthenticationStatus::kAuthenticated);
+}
+
+void EmailAliasesService::MaybeRequestSessionAgain() {
+  if (!session_poll_elapsed_timer_.has_value()) {
+    session_poll_elapsed_timer_.emplace();
+  }
+  if (session_poll_elapsed_timer_->Elapsed() > kMaxSessionPollDuration) {
+    LOG(ERROR) << "Email Aliases service verification error: exceeded max "
+                  "poll duration";
+    session_poll_elapsed_timer_.reset();
+    NotifyObserversAuthStateChanged(
+        mojom::AuthenticationStatus::kUnauthenticated,
+        /*error_message=*/l10n_util::GetStringUTF8(
+            IDS_EMAIL_ALIASES_ERROR_VERIFICATION_FAILED));
+    return;
+  }
+  // Schedule the next request after a short interval.
+  if (!session_request_timer_.IsRunning()) {
```

Comment:
It seems to me that if a call got here and this timer is running then something is wrong with the application logic....

Reply:
True. I changed it to a DCHECK.

Reply:
Actually to a CHECK.